lib/streamlit/ForwardMsgCache.py

```diff
@@ -96,19 +100,49 @@ class Entry(object):
         """
         def __init__(self, msg):
             self.msg = msg
-            self.sessions = WeakKeyDictionary()
+            # {ReportSession -> report_run_count}
+            self._session_report_run_counts = WeakKeyDictionary()
+
+        def add_session_ref(self, session, report_run_count):
+            """Adds a reference to a ReportSession that has referenced
+            this Entry's message.
+
+            Parameters
+            ----------
+            session : ReportSession
+            report_run_count : int
+                The session's run count at the time of the call
+
+            """
+            prev_run_count = self._session_report_run_counts.get(session, 0)
+            run_count = max(prev_run_count, report_run_count)
```

Comment:
Explain why this `max` is needed. I'm assuming it's to address a race condition.

Reply:
This is actually no longer needed, now that the message cache is only touched by one thread. I'll change it.